jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2024 Haulmont.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.jmix.flowui.data.binding.impl;
+
+import com.vaadin.flow.component.HtmlContainer;
+import com.vaadin.flow.shared.Registration;
+import io.jmix.core.MetadataTools;
+import io.jmix.core.common.datastruct.Pair;
+import io.jmix.core.common.event.Subscription;
+import io.jmix.core.common.util.Preconditions;
+import io.jmix.core.entity.EntitySystemAccess;
+import io.jmix.flowui.data.ValueSource;
+import io.jmix.flowui.data.binding.HtmlContainerReadonlyDataBinding;
+import io.jmix.flowui.kit.component.formatter.Formatter;
+import io.jmix.flowui.model.CollectionContainer;
+import io.jmix.flowui.model.InstanceContainer;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.lang.Nullable;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Component("flowui_HtmlContainerDataBinding")
+public class HtmlContainerReadonlyDataBindingImpl implements HtmlContainerReadonlyDataBinding {
+
+    protected MetadataTools metadataTools;
+
+    public HtmlContainerReadonlyDataBindingImpl(MetadataTools metadataTools) {
+        this.metadataTools = metadataTools;
+    }
+
+    protected List<Pair<HtmlContainer, HtmlContainerListenerRegistrationsCleaner>> listenersRegistrations = new LinkedList<>();
```

Comment:
More than 120 characters per line

List of a pairs... Maybe we need to use Map?